primitives/src/crh/poseidon/mod.rs

```diff
@@ -620,53 +520,93 @@ impl<F: PrimeField + MulShort, P: PoseidonParameters<Fr = F>> BatchFieldBasedHas
 
         // Assign pre-computed values of the state vector equivalent to a permutation with zero element state vector
-        let state_z = vec![P::AFTER_ZERO_PERM[0], P::AFTER_ZERO_PERM[1], P::AFTER_ZERO_PERM[2]];
+        //let state_z = vec![P::AFTER_ZERO_PERM[0], P::AFTER_ZERO_PERM[1], P::AFTER_ZERO_PERM[2]];
+        let mut state_z = Vec::new();
+        for i in 0..P::T {
+            state_z.push(P::AFTER_ZERO_PERM[i]);
+        }
 
         // Copy the result of the permutation to a vector of state vectors of the length equal to the length of the input
         // state is a vector of 3-element state vector.
         let mut state = Vec::new();
-        for _i in 0..input_length {
+        for _i in 0..array_length {
             state.push(state_z.clone());
         }
 
         // input_idx is to scan the input_array
         let mut input_idx = 0;
 
-        for k in 0..input_length {
-            state[k][0] += &input_array[input_idx];
-            input_idx += 1;
-            state[k][1] += &input_array[input_idx];
-            input_idx += 1;
-            // constant to add for a 2-1 Merkle tree
-            state[k][2] += &P::C2;
+        for k in 0..array_length {
+            for j in 0..P::R {
+                state[k][j] += &input_array[input_idx];
+                input_idx += 1;
+            }
+            // constant for m-ary Merkle tree
+            state[k][P::R] += &P::C2;
         }
 
         // apply permutation after adding the input vector
         Self::poseidon_perm_gen(&mut state);
 
-        // overwrite the input with the result of the hash
-        for k in 0..input_array.len()/2 {
-            input_array[k] = state[k][0];
+        // write the result of the hash extracted from the state vector to the output vector
+        for k in 0..array_length {
+            output_array[k] = state[k][0];
         }
     }
+
+    fn merkle_tree(input_vec: &mut[Self::Data], output_vec: &mut[Self::Data], input_size: usize){
+        // Supporting function that processes the inputs and outputs in chunks
+
+        let num_cores = 16;
+
+        assert_eq!(input_vec.len() % 2, 0, "The length of the input to the hash is not even.");
+        assert_eq!(output_vec.len() >= input_vec.len() / 2, true,  "The length of the output is not greater or equal to half of the input length.");
+
+        if input_size < 2 * num_cores {
+            input_vec.par_chunks_mut(2).zip(output_vec.par_chunks_mut(1)).for_each( |(p1,p2)| {
+                Self::batch_evaluate(p1, p2);
+            });
+            return;
+        }
+
+        use rayon::prelude::*;
+
+        input_vec.par_chunks_mut(input_size/num_cores).zip(output_vec.par_chunks_mut(input_size/num_cores/2)).for_each( |(p1, p2)| {
+            Self::batch_evaluate(p1, p2);
+        });
+
+    }
+
+    fn merkle_tree_2_1(array: &mut [Self::Data], input_size: usize) {
+        // Main function for the Merkle-tree computation
+
+        let mut copy_vec = &mut array[..];
```

Comment:
This would be unnecessary: you can declare "array" input param as mutable (mut array: &mut [Self::Data]) and avoid this